src/main.c

```diff
@@ -597,26 +611,47 @@ static void async_nb_do_async_transaction(ASYNC_JOB *job, neverbleed_iobuf_t *bu
 
 #endif
 
-static void async_nb_transaction(neverbleed_iobuf_t *buf)
+static void async_nb_transaction(neverbleed_iobuf_t *buf, int expect_response)
 {
 #if H2O_CAN_OSSL_ASYNC
     /* When using OpenSSL with ASYNC support, we may receive requests from a fiber. If so, process them asynchronously. */
     ASYNC_JOB *job;
-   if ((job = ASYNC_get_current_job()) != NULL) {
+    if ((job = ASYNC_get_current_job()) != NULL) {
+        assert(expect_response);
         assert(h2o_socket_get_fd(async_nb.sock) == neverbleed_get_fd(neverbleed));
         async_nb_do_async_transaction(job, buf);
         return;
     }
 #endif
 
-    /* do it synchronously, by at first reading all pending reads, then write asynchronously and wait for the repsonse */
-    if (async_nb.sock != NULL) {
-        while (async_nb.read_queue.len != 0)
-            async_nb_read_ready(async_nb.sock, NULL);
-    }
+    if (!expect_response && async_nb.sock != NULL) {
+        /* this is an optimization for fire-and-forget transactions which do not require a response.
+           these transactions are queued on the async socket with `expect_response` being false, are free'd
+           upon write completion in `async_nb_on_write_complete`
+        */
+        struct async_nb_transaction_t *transaction = h2o_mem_alloc(sizeof(*transaction));
+        *transaction = (struct async_nb_transaction_t){
+            .buf = h2o_mem_alloc(sizeof(*transaction->buf)),
+            .link = (h2o_linklist_t){},
+            .allocated = 1,
+            .expect_response = 0,
+        };
+        h2o_memcpy(transaction->buf, buf, sizeof(*buf));
```

Comment:
Do we want to "move" the object from `buf` to `transaction->buf` instead of creating a shallow copy, which blurs the ownership of the underlying pointer?

I mean, how about resetting `buf` to a zero buffer.